services/user-login-manager/app/userlogin.py

```diff
@@ -0,0 +1,237 @@
+# Copyright (C) 2023, BRAIN-LINK UG (haftungsbeschränkt). All Rights Reserved.
+# SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-ScanHub-Commercial
+
+"""Definition of the user management and login API endpoints (accessible through swagger UI)."""
+
+from typing import Annotated
+from hashlib import sha256, scrypt
+from secrets import compare_digest, token_hex
+from passlib.hash import argon2
+import time
+
+from fastapi import APIRouter, Depends, HTTPException, status
+from fastapi.security import OAuth2PasswordRequestForm
+
+from scanhub_libraries.models import User
+from scanhub_libraries.security import oauth2_scheme
+from app.db import UserSQL
+from app import dal
+
+
+AUTOMATIC_LOGOUT_TIME_SECONDS = 3600  # time until a login token gets invalid (1 hour)
+
+
+router = APIRouter()
+
+
+@router.get("/getcurrentuser", status_code=200, tags=["user"])
+async def get_current_user(access_token: Annotated[str, Depends(oauth2_scheme)]) -> User:
+    """Get current user endpoint.
+
+    Parameters
+    ----------
+    access_token 
+        User token (HTTP Header: "Authorization: Bearer <access_token>")
+
+    Returns
+    -------
+        User pydantic model, the user data of the current user.
+
+    Raises
+    ------
+    HTTPException
+        401: Unauthorized if the token is invalid.
+    """
+    user_db: UserSQL = await dal.get_user_from_token(access_token)
+    if user_db is not None and user_db.access_token is not None and user_db.last_activity_unixtime is not None \
+            and time.time() - user_db.last_activity_unixtime < AUTOMATIC_LOGOUT_TIME_SECONDS:              # auto-logout time 1 hour
+        await dal.update_user_data(user_db.username, {"last_activity_unixtime": time.time()})         # reset last_activity timer
+        user = await get_user_out(user_db)
+        return user
+
+    print("Received invalid token.")
+    raise HTTPException(
+        status_code=status.HTTP_401_UNAUTHORIZED,
+        detail="Invalid token.",
+        headers={"WWW-Authenticate": "Bearer"},
+    )
+
+
+def compute_complex_password_hash(password: str, salt: str) -> str:
```

Comment:
salted sha256 is not an ideal choice anymore

Reply:
It is not just a salted sha256, but a series of 
1. sha256 (without salt)
2. scrypt (with salt) (key derivation function from python standard library)
3. argon2 (with salt) (more modern key derivation function from passlib)
4. sha256 (without salt)